From b171e096b5506694744ec16155e053c03fb16456 Mon Sep 17 00:00:00 2001
From: Iain Sandoe <iain@sandoe.co.uk>
Date: Wed, 2 Sep 2020 20:33:00 +0100
Subject: [PATCH] Darwin : Make trampoline templates linker-visible.

For Arm64, the alignment of the LTRAMPn symbols matters.

Actually, the LTRAMPn  symbols _are_ 8 byte aligned, but because
they are Local, the linker doesn't know that this guarantee can be met.
It assumes that they are not necessarily more aligned than the
containing section (ld64 atoms strike again).

The fix is to publish the trampoline symbol for the linker to access
directly - it can then see that the atom is suitably aligned.

Fixes issue #11.

(cherry picked from commit 717057366d780cc3f54d5621e78334a3e63789d0)
---
 gcc/config/darwin.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git gcc/config/darwin.h gcc/config/darwin.h
index e360e771c6d..076ff9f29eb 100644
--- gcc/config/darwin.h
+++ gcc/config/darwin.h
@@ -895,6 +895,8 @@ extern GTY(()) section * darwin_sections[NUM_DARWIN_SECTIONS];
       sprintf (LABEL, "*%s%ld", "lubsan_type", (long)(NUM));\
     else if (strcmp ("LASAN", PREFIX) == 0)	\
       sprintf (LABEL, "*%s%ld", "lASAN", (long)(NUM));\
+    else if (strcmp ("LTRAMP", PREFIX) == 0)	\
+      sprintf (LABEL, "*%s%ld", "lTRAMP", (long)(NUM));\
     else						\
       sprintf (LABEL, "*%s%ld", PREFIX, (long)(NUM));	\
   } while (0)
-- 
2.40.1

